hashintel/hash
# H-1587, H-1949, H-2103: Better type permission checks, hide warning when creating link types

## 🌟 What is the purpose of this PR?

Makes the following type permission-related changes:
1. Allow checking a user's permissions on a type via the API
2. On the new entity page, show an error message if a user can't create entities of that type
3. Hide the 'create entity' button on the type page if a user can't create entities of that type
4. On the type page, ese the permissions from the API to check if the type form should be readonly or not, rather than members of the org can edit any type in the org

Smaller drive-bys:
- fixes H-1949 by hiding the "You are extending a type" warning when creating a new link type (all link types extend the Block Protocol link type)
- catch errors when trying to create an entity and show a warning modal
- hide 'Profile Bio' and 'Text' entities from the sidebar entities list

Following further testing:
- Fix query for entities page (H-2103) and add test
- Don't show 'Loading...' for entities table with no results

## Pre-Merge Checklist 🚀

### 🚢 Has this modified a publishable library?




This PR:


### 📜 Does this require a change to the docs?




The changes in this PR:


### 🕸️ Does this require a change to the Turbo Graph?




The changes in this PR:


## ⚠️ Known issues

It would be better to hide types that users can't create from the 'select type of entity to create' dropdown, and also when selecting parents of a type, but this is a much bigger change since it means updating how we fetch all entity types to also fetch their permissions. Created H-2027 to address this later.



## 🛡 What tests cover this?



- Type creation is covered by a Playwright test

## ❓ How to test this?



1. Check that you can still types you're supposed to be able to and can't edit types you're not
2. Visit type page for Actor, check 'create entity' button doesn't show
3. Try creating a new entity of type Actor (view /new/entity route)

## 📹 Demo



https://github.com/hashintel/hash/assets/37743469/13c5b235-e1d1-4fe3-bd55-3b3339864816
